### PR TITLE
Add validation to ssh -X scenario

### PR DESCRIPTION
**Pull request by JRivrain**
Related ticket https://progress.opensuse.org/issues/80780

As of now we only check that system boots, but do not even verify that ssh connection is available after the initial reboot.

We already have modules to validate ssh service is running and port is open.
Check tests/console/sshd.pm and tests/console/sshd_running.pm
However, one only checks if the service is up the other one shuts the firewall down, so we need a new module, or to modify the existing ones.
I chose to create a new one as those modules are used by different teams, and the sshd test is overkill for what we want here.

VRs:
PVM https://openqa.suse.de/tests/5249481
zVM https://openqa.suse.de/tests/5249482
remote_ssh_target_ftp https://openqa.suse.de/tests/5249485

### CODE REVIEW

**Review by jknphy — COMMENTED** [review]
I just reviewed your code before checking existing modules, now I see that you borrow that code, we don't need a new module, we probably need a library with shared functions if there are parts that do not works and we don't know why, but first, could you please share the problems with sshd if it is the case?

File: tests/installation/validation/validate_sshd_reachable.pm [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+use strict;
+use warnings;
+use base "opensusebasetest";
+use testapi;
+use utils "systemctl";
+
+sub run {
+    select_console 'root-console';
+    systemctl 'show -p ActiveState sshd|grep ActiveState=active';
```
[the remark this answers is not in the document]
Comment (JRivrain) [reply tests/installation/validation/validate_sshd_reachable.pm:22]: I know, I copied the code from the already existing and tested test module "sshd". Basically this new module is just code copied from sshd and sshd_running.

File: tests/installation/validation/validate_sshd_reachable.pm [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# SUSE's openQA tests
+#
+# Copyright © 2020 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved. This file is offered as-is,
+# without any warranty.
+#
+# Summary: Validate the sshd is running *and* that port 22 is opened,as
```
[the remark this answers is not in the document]
Comment (JRivrain) [reply tests/installation/validation/validate_sshd_reachable.pm:10]: I think it helps to differentiate with sshd and sshd_running modules

**Comment by JRivrain** [comment]
> I just reviewed your code before checking existing modules, now I see that you borrow that code, we don't need a new module, we probably need a library with shared functions if there are parts that do not works and we don't know why, but first, could you please share the problems with sshd if it is the case?

it shuts the firewall down. its purpose is to check extensively the daemon itself, not to verify that ssh is actually available *from outside*.

**Comment by jknphy** [comment]
> > I just reviewed your code before checking existing modules, now I see that you borrow that code, we don't need a new module, we probably need a library with shared functions if there are parts that do not works and we don't know why, but first, could you please share the problems with sshd if it is the case?
> 
> it shuts the firewall down. its purpose is to check extensively the daemon itself, not to verify that ssh is actually available _from outside_.

yes, but the idea was to reuse something, I still don't see you showed the errors you found there, for what I can hear, we could just use it and put the firewalld up after the test (even not needed for now, because we might not schedule anything after this module).
The new test that you added doesn't test from outside neither, it is just an extraction of the basics things, and should be shared in some library not repeated again. We don't even test anything about X forwarding if we are trying to be that specific.
In summary, I can see two options, (1) we can go with new test but we should find for more things specific and use shared libraries (2) use existing tests. wdyt?

**Comment by JRivrain** [comment]
> > > I just reviewed your code before checking existing modules, now I see that you borrow that code, we don't need a new module, we probably need a library with shared functions if there are parts that do not works and we don't know why, but first, could you please share the problems with sshd if it is the case?
> > 
> > 
> > it shuts the firewall down. its purpose is to check extensively the daemon itself, not to verify that ssh is actually available _from outside_.
> 
> yes, but the idea was to reuse something, I still don't see you showed the errors you found there, 

There are no errors. Those modules are just not adapted for what we want to test.

> for what I can hear, we could just use it and put the firewalld up after the test 

Again, the purpose of the test is to probe that the ssh server is accessible *from outside* and *right after install*. Meaning, with the firewall up.

> ... That module seems to test in deep the server.

Yes. This is not what we need, it's out of scope for us.

> The new test that you added doesn't test from outside neither, it is just an extraction of the basics things, and should be shared in some library not repeated again. We don't even test anything about X forwarding if we are trying to be that specific.

The test module does not aim to test X forwarding, but to test that an adming can connect to the machine after install.

> In summary, I can see two options, (1) we can go with new test but we should find for more things specific and use shared libraries (2) use existing tests if they do not fail and add new test module for anything additional. wdyt?

I can put those in shared libraries. I just wanted to avoid touching other team's modules. Other than that, I don't share your opinions.

**Comment by JRivrain** [comment]
Another option is to an an "if (SOMEVARIABLE) { don't shutdown the firewall }" in sshd. Though, the test may need this firewall down to work properly. This is why I did not go for that option at the first place, when it is so simple to create a new module that does exactly what we need, and nothing more.

**Comment by jknphy** [comment]
> Again, the purpose of the test is to probe that the ssh server is accessible _from outside_ and _right after install_. Meaning, with the firewall up.

Leave it like that without shared libraries, the command are simple enough, I thought we could reuse more from those tests and you found some issue that you didn't mentioned, but it is not the case, your explanation about 'from outside' confused me because both are not testing anything from outside, one is more simple that the other and does not require the firewall down, because does not try connection client/server in localhost.

My bad about the X11 fw, I thought that the original idea of the ticket was to check X11 forwarding, but it is not mentioned in the ticked explicitly, i found it in other bugs so I guess it is optional, I found interesting for example to check in `/etc/ssh/sshd_config` X11Forwarding Yes. Do you think that we could check anything else there?

**Comment by jknphy** [comment]
> > > > I just reviewed your code before checking existing modules, now I see that you borrow that code, we don't need a new module, we probably need a library with shared functions if there are parts that do not works and we don't know why, but first, could you please share the problems with sshd if it is the case?
> > > 
> > > 
> > > it shuts the firewall down. its purpose is to check extensively the daemon itself, not to verify that ssh is actually available _from outside_.
> > 
> > 
> > yes, but the idea was to reuse something, I still don't see you showed the errors you found there,
> 
> There are no errors. Those modules are just not adapted for what we want to test.
> 
> > for what I can hear, we could just use it and put the firewalld up after the test
> 
> Again, the purpose of the test is to probe that the ssh server is accessible _from outside_ and _right after install_. Meaning, with the firewall up.
> 
> > ... That module seems to test in deep the server.
> 
> Yes. This is not what we need, it's out of scope for us.
> 
> > The new test that you added doesn't test from outside neither, it is just an extraction of the basics things, and should be shared in some library not repeated again. We don't even test anything about X forwarding if we are trying to be that specific.
> 
> The test module does not aim to test X forwarding, but to test that an admin can connect to the machine after install. There is no possibility to have MM test on those backends.
> 
> > In summary, I can see two options, (1) we can go with new test but we shoud find for more things specific and use shared libraries (2) use existing tests if they do not fail and add new test module for anything additional. wdyt?
> 
> I can put those in shared libraries. I just wanted to avoid touching other team's modules. Other than that, I don't share your opinions.

I think we are almost in the same page, just there was some misunderstanding :)

**Comment by JRivrain** [comment]
> > Again, the purpose of the test is to probe that the ssh server is accessible _from outside_ and _right after install_. Meaning, with the firewall up.
> 
> I thought we could reuse more from those tests and you found some issue that you didn't mentioned, but it is not the case, your explanation about 'from outside' confused me because both are not testing anything from outside, one is more simple that the other and does not require the firewall down, because does not try connection client/server in localhost.
> 
> My bad about the X11 fw, I thought that the original idea of the ticket was to check X11 forwarding, but it is not mentioned in the ticked explicitly, i found it in other bugs so I guess it is optional, I found interesting for example to check in `/etc/ssh/sshd_config` X11Forwarding Yes. Do you think that we could check anything else there?

Ah, in the meantime I added this to a shared lib :-|. Well yes, that could be a good extra check, but I'm not sure if it's actually a requirement in SLE ?

**Comment by jknphy** [comment]
> Another option is to an an "if (SOMEVARIABLE) { don't shutdown the firewall }" in sshd. Though, the rest of test may need this firewall down to work properly. This is why I did not go for that option at the first place, when it is so simple to create a new module that does exactly what we need, and nothing more.

you are thinking in the good way, just I thought we could extend a bit more the scenario once we saw that the extracted part is small. That option with the setting better not, probably works only with firewalld down.

**Comment by jknphy** [comment]
> > > Again, the purpose of the test is to probe that the ssh server is accessible _from outside_ and _right after install_. Meaning, with the firewall up.
> > 
> > 
> > I thought we could reuse more from those tests and you found some issue that you didn't mentioned, but it is not the case, your explanation about 'from outside' confused me because both are not testing anything from outside, one is more simple that the other and does not require the firewall down, because does not try connection client/server in localhost.
> > My bad about the X11 fw, I thought that the original idea of the ticket was to check X11 forwarding, but it is not mentioned in the ticked explicitly, i found it in other bugs so I guess it is optional, I found interesting for example to check in `/etc/ssh/sshd_config` X11Forwarding Yes. Do you think that we could check anything else there?
> 
> Ah, in the meantime I added this to a shared lib :-|. Well yes, that could be a good extra check, but I'm not sure if it's actually a requirement in SLE ?

Only YaST developers could know for sure in case you want to check it, Idk, probably in this scenario is safe to add it, there is another one in `/etc/ssh_config` but in other scenarios might not be default.

File: lib/services/sshd.pm [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+package services::sshd;
+use base 'opensusebasetest';
+use testapi;
+use utils;
+use strict;
+use warnings;
+
+
+sub check_sshd_ports_opened {
```
Comment (jknphy) [inline lib/services/sshd.pm:21]: both protocols use the same port, could we find another name? for example `check_sshd_port`?

File: lib/services/sshd.pm [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# SUSE's openQA tests
+#
+# Copyright © 2019 SUSE LLC
```
Comment (jknphy) [inline lib/services/sshd.pm:3]: copyright date

File: tests/installation/validation/validate_sshd_reachable.pm [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+# admins may want to connect with ssh right after an ssh installation.
+# Maintainer: QE YaST <qa-sle-yast@suse.de>
+
+use strict;
+use warnings;
+use base "opensusebasetest";
+use testapi;
```
Comment (jknphy) [inline tests/installation/validation/validate_sshd_reachable.pm:17]: not needed.

File: lib/services/sshd.pm (hunk 3 above)
Comment (JRivrain) [reply lib/services/sshd.pm:21]: makes sense

File: lib/services/sshd.pm (hunk 4 above)
Comment (JRivrain) [reply lib/services/sshd.pm:3]: > > Again, the purpose of the test is to probe that the ssh server is accessible _from outside_ and _right after install_. Meaning, with the firewall up.
> 
> I think we are almost in the same page, just there was some misunderstanding :)
> 
> I thought we could reuse more from those tests and you found some issue that you didn't mentioned, but it is not the case, your explanation about 'from outside' confused me because both are not testing anything from outside, one is more simple that the other and does not require the firewall down, because does not try connection client/server in localhost.
> 
> My bad about the X11 fw, I thought that the original idea of the ticket was to check X11 forwarding, but it is not mentioned in the ticked explicitly, i found it in other bugs so I guess it is optional, I found interesting for example to check in `/etc/ssh/sshd_config` X11Forwarding Yes. Do you think that we could check anything else there?

I guess I could add something like "ssh root@localhost 'true'", which would return 0 if the connexion works.

**Comment by JRivrain** [comment]
> Do you think that we could check anything else there?

I guess I could add something like "ssh root@localhost 'true'", which would return 0 if the connexion works.

**Comment by jknphy** [comment]
> > Do you think that we could check anything else there?
> 
> I guess I could add something like "ssh root@localhost 'true'", which would return 0 if the connexion works.

My preference in my experience is always to check configuration, do actions for real always brings some headache of maintenance, but it that works, is simple enough.

**Comment by JRivrain** [comment]
> > > Do you think that we could check anything else there?
> > 
> > 
> > I guess I could add something like "ssh root@localhost 'true'", which would return 0 if the connexion works.
> 
> My preference in my experience is always to check configuration, do actions for real always brings some headache of maintenance (plus is already tested upstream), but it that works, is simple enough.
> 
> For settings you can check: https://en.opensuse.org/SDB:Configure_openSSH YaST should be doing something really similar, for instance, uncommented seems to be (in my local) `ForwardX11Trusted yes` and `Host *`

Effectively it's not that simple to connect with ssh locally because of password prompt, so we'd need to use expect or some dirty "script_run". I don't like the idea to probe for X11 forwarding after all, as this could never be considered as a bug if this is not enabled, and it's easy to enable it if we have an ssh access. Even if the user absolutely wants to do it in a UI, he can still use ncurses. But it looks like a good idea to check that we have at least "Host *" in the config file.

**Review by jknphy — APPROVED** [review]
LGTM, but before merging, please update copyright date in all files.
